### PR TITLE
fix json log format panic, change the flag name in flagIsSet

### LINKED ISSUE #99239: Kubelet panics when trying to set structured logging
#### What happened:
When I attempt to start the kubelet with the structured logging flag (--logging-format=json) the kubelet panics with a `panic: failed to lookup unsupported log flag` error.

#### What you expected to happen:
I expect the Kubelet to start with structured logging enabled.

#### How to reproduce it (as minimally and precisely as possible):
```
KUBELET_FLAGS=--logging-format=json hack/local-up-cluster.sh
```

Adding the following yaml within hack/local-up-cluster.sh for the kubelet config fails in the same manner:
```
logging:
  format: json
```

#### Environment:
- Kubernetes version (use `kubectl version`): master 25aa6cf0c58067656dd6e4c1e2c28efb259ab08c
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by chenyw1990**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
kubelet will panic with --logging-format=json

#### Which issue(s) this PR fixes:

Fixes #99239

#### Special notes for your reviewer:

@serathius
@ehashman
@knight42 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by chenyw1990** [comment]
retest

**Comment by serathius** [comment]
/hold 
This is not correct way to fix the issue. Json format doesn't support all klog flags, this logic is meant to protect against trying to enable JSON with one of the unsupported flags, so the logic is correct.

Looks like new version of klog brought new flag `add-dir-header` that somehow is not found by `flagIsSet` function meant to check flag is set.

**Comment by serathius** [comment]
https://github.com/kubernetes/kubernetes/pull/98524 looks like a culprit.
It checked names of flags that are returned by `unsupportedLoggingFlags` without updating `flagIsSet` function.

**Comment by serathius** [comment]
Proposed fix, instead of adding `json` in `Validate` function, can you change `flagIsSet` function to
```
func flagIsSet(name string) bool {
	f := flag.Lookup(name)
	if f != nil {
		return f.DefValue != f.Value.String()
	}
	f = flag.Lookup(strings.Replace(name, "-", "_", -1))
	if f != nil {
		return f.DefValue != f.Value.String()
	}
	pf := pflag.Lookup(name)
	if pf != nil {
		return pf.DefValue != pf.Value.String()
	}
	panic(fmt.Sprintf("failed to lookup unsupported log flag: %s", name))
}
```

**Comment by chenyw1990** [comment]
> Proposed fix, instead of adding `json` in `Validate` function, can you change `flagIsSet` function to
> 
> ```
> func flagIsSet(name string) bool {
> 	f := flag.Lookup(name)
> 	if f != nil {
> 		return f.DefValue != f.Value.String()
> 	}
> 	f = flag.Lookup(strings.Replace(name, "-", "_", -1))
> 	if f != nil {
> 		return f.DefValue != f.Value.String()
> 	}
> 	pf := pflag.Lookup(name)
> 	if pf != nil {
> 		return pf.DefValue != pf.Value.String()
> 	}
> 	panic(fmt.Sprintf("failed to lookup unsupported log flag: %s", name))
> }
> ```

sure, I have modify the code.

**Comment by serathius** [comment]
/priority critical-urgent
If not merged, 1.21 will have json logging broken.

File: staging/src/k8s.io/component-base/logs/options.go [hunk 1]
```diff
@@ -75,6 +75,10 @@ func flagIsSet(name string) bool {
 	if f != nil {
 		return f.DefValue != f.Value.String()
 	}
+	f = flag.Lookup(strings.Replace(name, "-", "_", -1))
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/options.go:78]: can we use 

doing this transformation here seems fragile/wrong... does this mean the transformation in unsupportedLoggingFlags() is incorrect?

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: It's reverse of one in `unsupportedLoggingFlags`.
One of the places we do this normalization https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/component-base/cli/globalflag/globalflags.go#L52

Comment (liggitt) [reply staging/src/k8s.io/component-base/logs/options.go:78]: > It's reverse of one in unsupportedLoggingFlags.

right... special-casing this to undo a transformation done there seems fragile. It seems like the caller of this function should pass the correct flag name, rather than making this function transform and fallback...

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: Yes, I didn't prioritize it thou as we are also working on unifying klog flag setup. https://github.com/kubernetes/kubernetes/issues/99265

Possibly we should take one step ahead, wrap klog by `component-base/logs` and forbid direct import of `klog` library. That way all the klog flags will be controlled in one place in k/k. So we avoid fragile copying. This is bigger effort thou

Comment (ehashman) [reply staging/src/k8s.io/component-base/logs/options.go:78]: I don't understand why we need to do the normalization at all, this seems to be a new breaking change in 1.21. Can we just leave the klogs flags alone and revert all these `_ -> -` changes? Personally I think we should have rejected https://github.com/kubernetes/kubernetes/pull/98524 as it is an `ACTION REQUIRED:` end user change for the (minimal) benefit of flag format consistency.

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: #98524 didn't introduce any breaking changes. It's goal was to fix inconsistency between help message in --logging-format command and flags registered in components.

Comment (ehashman) [reply staging/src/k8s.io/component-base/logs/options.go:78]: Where are they getting normalized then? How did a help text fix break a flag? This is why I am confounded.

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: They are normalized here https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/component-base/cli/globalflag/globalflags.go#L45
Assuming that components calls globalflags.AddGlobalFlags instead of klog.Init

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: As for why we need to both handle names with and without normalization, reason is that depending on how klog is initialized those flags will be registered under different names.

**Comment by ehashman** [comment]
There are at least 3 other components that aren't the kubelet using this config flag, have we verified if any of them are affected? Since we're making a component base change here we should probably also add tests for the other consumers, I'm not sure if they also panic on log-format=json.

**Comment by serathius** [comment]
I verified that scheduler panics without this change.

File: staging/src/k8s.io/component-base/logs/options.go (hunk 1 above)
Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: @liggitt Is the proposed approach for a fix acceptable? If we would want to remove fragile string replace logic before code freeze we should start working on it early.

Comment (liggitt) [reply staging/src/k8s.io/component-base/logs/options.go:78]: I'm not seeing why this validation method needs to look up flags at all... can it do this and sidestep the whole issue?

```
func (o *Options) Validate() []error {
	errs := []error{}
	if _, err := o.Get(); err != nil {
		errs = append(errs, fmt.Errorf("unsupported log format: %s", o.LogFormat))
	} else if o.LogFormat != defaultLogFormat {
		errs = append(errs, fmt.Errorf("non-default logging format %q is not supported", o.LogFormat))
	}
	return errs
}
```

Comment (serathius) [reply staging/src/k8s.io/component-base/logs/options.go:78]: It looks up flags because json format doesn't support all the flags that klog implements. When adding json format I was requested to protect users from trying to use klog specific features with json logging.

Comment (liggitt) [reply staging/src/k8s.io/component-base/logs/options.go:78]: > It looks up flags because json format doesn't support all the flags that klog implements. When adding json format I was requested to protect users from trying to use klog specific features with json logging.

I see, I misunderstood the flag/format interaction. Would something like this work instead:

```go
	if o.LogFormat != defaultLogFormat {
		allFlags := unsupportedLoggingFlags()
		for _, fname := range allFlags {
			f := flag.Lookup(fname)
			pf := pflag.Lookup(fname)
			if (f != nil && f.DefValue != f.Value.String()) || (pf != nil && pf.DefValue != pf.Value.String()) {
				errs = append(errs, fmt.Errorf("non-default logging format %q doesn't honor flag: %s", o.LogFormat, fname))
			}
		}
	}
```

Comment (liggitt) [reply staging/src/k8s.io/component-base/logs/options.go:78]: > It looks up flags because json format doesn't support all the flags that klog implements. When adding json format I was requested to protect users from trying to use klog specific features with json logging.

I see, I misunderstood the flag/format interaction. Yes, we have to ensure unsupported options are not set incorrectly and ignored.

> @liggitt Is the proposed approach for a fix acceptable? If we would want to remove fragile string replace logic before code freeze we should start working on it early.

I don't think it is acceptable. This is already really hard to follow, I don't think we should make it worse, making assumptions about the global pflag normalization functions and undoing changes we made in `unsupportedLoggingFlags`.

I think `unsupportedLoggingFlags` should probably be told how to normalize flags, and the two callers can pass in the normalization function for their respective flagsets, rather than hard-coding `_` and `-` replacements.

Comment (chenyw1990) [reply staging/src/k8s.io/component-base/logs/options.go]: @liggitt @serathius I have modified the code based on the review comments.

File: staging/src/k8s.io/component-base/logs/options.go [hunk 2]
```diff
@@ -70,11 +70,29 @@ func (o *Options) Validate() []error {
...
+// we should always use hyphens instead of underscores when registering component flags
+func Normalize(s string) string {
+	return strings.Replace(s, "_", "-", -1)
+}
+
+// UnNormalize replaces hyphens with underscores
+// we should always use underscores instead of hyphens when validate flags
+func UnNormalize(s string) string {
+	return strings.Replace(s, "-", "_", -1)
+}
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/options.go:83]: pflag flagsets already have normalization functions (`GetNormalizeFunc()`)... we should use those instead of defining unrelated ones here

**Comment by ehashman** [comment]
@chenyw1990 PTAL at the failing unit test, it looks related:

```
 === FAIL: cmd/kube-apiserver/app/options TestAddCustomGlobalFlags (0.00s)
    globalflags_test.go:59: [Default]: expected [add-dir-header alsologtostderr cloud-provider-gce-l7lb-src-cidrs cloud-provider-gce-lb-src-cidrs default-not-ready-toleration-seconds default-unreachable-toleration-seconds help log-backtrace-at log-dir log-file log-file-max-size log-flush-frequency logtostderr one-output skip-headers skip-log-headers stderrthreshold v vmodule], got [add_dir_header alsologtostderr cloud-provider-gce-l7lb-src-cidrs cloud-provider-gce-lb-src-cidrs default-not-ready-toleration-seconds default-unreachable-toleration-seconds help log-flush-frequency log_backtrace_at log_dir log_file log_file_max_size logtostderr one_output skip_headers skip_log_headers stderrthreshold v vmodule] 
```

**Comment by ehashman** [comment]
I think all comments should be addressed?

@liggitt would you PTAL?

File: staging/src/k8s.io/component-base/logs/options.go [hunk 3]
```diff
@@ -70,11 +70,23 @@ func (o *Options) Validate() []error {
 	return errs
 }
 
-func flagIsSet(name string) bool {
+// UnNormalize replaces hyphens with underscores
+// we should always use underscores instead of hyphens when validate flags
+func UnNormalize(s string) string {
+	return strings.Replace(s, "-", "_", -1)
+}
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/options.go:77]: since this is only used locally, don't export it

I'd also suggest a more descriptive name like `hyphensToUnderscores`

**Review by liggitt — COMMENTED** [review]
This looks a lot more coherent, thanks. Just had a couple suggestions about the unnormalize func

/approve
/hold for UnNormalize changes
instrumentation reviewer has lgtm once that is addressed

**Comment by chenyw1990** [comment]
@liggitt Thanks for your suggestions, I have modified the code based on the review comments.

**Comment by serathius** [comment]
/hold cancel
as UnNormalize changes were addressed

**Comment by rifelpet** [comment]
@chenyw1990 We're seeing failures in the kOps e2e jobs using CI builds of k8s due to the `--log-file` flag changing to `--log_file`

```
$ docker run -it --rm k8s.gcr.io/kube-apiserver:v1.20.4 -- kube-apiserver --help
...
      --log-file string                  If non-empty, use this log file
      --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
```

however recent CI builds show
```
$ curl https://storage.googleapis.com/kubernetes-release-dev/ci/v1.21.0-beta.1.200+2eb32e930a30ec/bin/linux/amd64/kube-apiserver.tar | docker load
$ docker run -it --rm k8s.gcr.io/kube-apiserver-amd64:v1.21.0-beta.1.200_2eb32e930a30ec -- kube-apiserver --help
      --log_file string                  If non-empty, use this log file
      --log_file_max_size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
```


Is this intentional?

**Comment by chenyw1990** [comment]
@rifelpet sorry,  I have add a new PR to fixed this change. #100171
